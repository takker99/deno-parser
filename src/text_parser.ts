import { makeExec, makeTryExec, type ParseFinalResult } from "./exec.ts";
import type { BaseLocation, ReaderTuple } from "./reader.ts";
import type { Parser } from "./parser.ts";

export type TextPosition = readonly [
  index: number,
  line: number,
  column: number,
];
export type TextSeeker = readonly [
  currentPosition: TextPosition,
  stack: readonly TextPosition[],
];
export interface TextLocation extends BaseLocation {
  line: number;
  column: number;
}
export interface TextReader {
  position: TextPosition;
  input: string;
  seeker: TextSeeker;
  location: TextLocation;
}
const initialSeeker: TextSeeker = [[0, 1, 1], []];
const textReader: ReaderTuple<TextReader> = [
  initialSeeker,
  (context, size) => {
    if (size == 0) return [false, context, ""];
    const [input, seeker] = context;
    if (size < 0 || input == "") return [true, context];
    let [[index, line, column], stack] = seeker ?? initialSeeker;
    const sliced = input.slice(index, index + size);
    const len = sliced.length;
    if (len == 0) return [true, context];
    index += len;
    for (const char of sliced) {
      if (char == "\n") {
        line++;
        column = 1;
      } else {
        column++;
      }
    }
    return [false, [input, [[index, line, column], stack]], sliced];
  },
  ([cur, stack]) => [cur, [cur, ...stack]],
  (seeker) => {
    if (seeker[1].length === 0) return seeker;
    const [, [prev, ...stack]] = seeker;
    return [prev, stack];
  },
  (seeker) => {
    if (seeker[1].length === 0) return seeker;
    const [cur, [, ...stack]] = seeker;
    return [cur, stack];
  },
  ([input, seeker]) => (seeker?.[0]?.[0] ?? 0) >= input.length,
  (a, b) => a[0][0] - b[0][0],
  ([[index, line, column]]) => ({ index, line, column }),
];

export const parse: <A>(
  parser: Parser<A, TextReader>,
  input: string,
) => ParseFinalResult<A, TextLocation> = /* #__PURE__ */ makeExec(textReader);

export const tryParse: <A>(parser: Parser<A, TextReader>, input: string) => A =
  /* #__PURE__ */ makeTryExec(textReader);
