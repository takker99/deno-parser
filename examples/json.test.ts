import { assertEquals } from "@std/assert";
import { parse } from "../src/text_parser.ts";
import { JSON } from "./json.ts";

Deno.test("json complex", () =>
  assertEquals(
    parse(
      JSON,
      `\
{
  "id": "a thing\\nice\tab",
  "another property!"
    : "also cool"
  , "weird formatting is ok too........ðŸ˜‚": 123.45e1,
  "": [
    true, false, null,
    "",
    " ",
    {},
    {"": {}}
  ]
}
`,
    ),
    {
      ok: true,
      value: {
        // deno-fmt-ignore
        "": [ true, false, null, "", " ", {}, { "": {}, }, ],
        "another property!": "also cool",
        "weird formatting is ok too........ðŸ˜‚": 1234.5,
        id: `a thing
ice\tab`,
      },
    },
  ));

Deno.test("json simple", () =>
  assertEquals(
    parse(JSON, `{"array":[1,"two",null,true,false],"obj":{}}`),
    {
      ok: true,
      value: {
        array: [1, "two", null, true, false],
        obj: {},
      },
    },
  ));

Deno.test("json multiline", () =>
  assertEquals(
    parse(
      JSON,
      `
{
  "array": [1, "two", null, true, false]
  "obj": {}
}
`,
    ),
    {
      expected: [",", "}"],
      location: { column: 3, index: 46, line: 4 },
      ok: false,
    },
  ));

Deno.test("json multiline extra weird", () =>
  assertEquals(
    parse(
      JSON,
      `
{
  "array" : [ 1 , "two", null , true , false ] ,
  "obj" : {


        "key"       : "value"


  }
}
`,
    ),
    {
      ok: true,
      value: {
        array: [1, "two", null, true, false],
        obj: { key: "value" },
      },
    },
  ));

Deno.test("json unicode escape", () =>
  assertEquals(parse(JSON, `{ "space": "\\u0020" }`), {
    ok: true,
    value: { space: " " },
  }));

Deno.test("json named escape", () =>
  assertEquals(parse(JSON, `{ "bnfrt": "\\b\\n\\f\\r\\t" }`), {
    ok: true,
    value: {
      bnfrt: `
\r	`,
    },
  }));

Deno.test("json useless escape", () =>
  assertEquals(parse(JSON, `{ "xyz": "\\x\\y\\z" }`), {
    ok: true,
    value: { xyz: "xyz" },
  }));
