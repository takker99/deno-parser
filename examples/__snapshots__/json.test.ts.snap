export const snapshot = {};

snapshot[`json complex 1`] = `
{
  ok: true,
  value: {
    "": [
      true,
      false,
      null,
      "",
      " ",
      {},
      {
        "": {},
      },
    ],
    "another property!": "also cool",
    "weird formatting is ok too........ðŸ˜‚": 1234.5,
    id: "a thing
ice	ab",
  },
}
`;

snapshot[`json simple 1`] = `
{
  ok: true,
  value: {
    array: [
      1,
      "two",
      null,
      true,
      false,
    ],
    obj: {},
  },
}
`;

snapshot[`json multiline 1`] = `
{
  expected: [
    ",",
    "}",
  ],
  location: {
    column: 3,
    index: 46,
    line: 4,
  },
  ok: false,
}
`;

snapshot[`json multiline extra weird 1`] = `
{
  ok: true,
  value: {
    array: [
      1,
      "two",
      null,
      true,
      false,
    ],
    obj: {
      key: "value",
    },
  },
}
`;

snapshot[`json unicode escape 1`] = `
{
  ok: true,
  value: {
    space: " ",
  },
}
`;

snapshot[`json named escape 1`] = `
{
  ok: true,
  value: {
    bnfrt: "
\\r	",
  },
}
`;

snapshot[`json useless escape 1`] = `
{
  ok: true,
  value: {
    xyz: "xyz",
  },
}
`;
